dev-python/mypy/mypy-1.5.1-r1.ebuild

```diff
@@ -0,0 +1,77 @@
+# Copyright 1999-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DISTUTILS_EXT=1
+DISTUTILS_USE_PEP517=setuptools
+PYTHON_COMPAT=( python3_{10..12} )
+
+inherit distutils-r1 multiprocessing
+
+DESCRIPTION="Optional static typing for Python"
+HOMEPAGE="
+	https://www.mypy-lang.org/
+	https://github.com/python/mypy/
+	https://pypi.org/project/mypy/
+"
+SRC_URI="
+	https://github.com/python/mypy/archive/v${PV}.tar.gz
+		-> ${P}.gh.tar.gz
+"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~loong ~mips ~ppc ~ppc64 ~riscv ~s390 ~sparc ~x86"
+IUSE="+native-extensions"
+
+# stubgen collides with this package: https://bugs.gentoo.org/585594
+RDEPEND="
+	!dev-util/stubgen
+	>=dev-python/psutil-4[${PYTHON_USEDEP}]
+	>=dev-python/typing-extensions-4.1.0[${PYTHON_USEDEP}]
+	>=dev-python/mypy_extensions-1.0.0[${PYTHON_USEDEP}]
+	$(python_gen_cond_dep '
+		dev-python/tomli[${PYTHON_USEDEP}]
+	' 3.{9..10})
+"
+BDEPEND="
+	native-extensions? (
+		dev-python/types-psutil[${PYTHON_USEDEP}]
+		dev-python/types-setuptools[${PYTHON_USEDEP}]
+	)
+	test? (
+		>=dev-python/attrs-18.0[${PYTHON_USEDEP}]
+		>=dev-python/filelock-3.3.0[${PYTHON_USEDEP}]
+		>=dev-python/lxml-4.4.0[${PYTHON_USEDEP}]
+		>=dev-python/pytest-7.4.0[${PYTHON_USEDEP}]
+		>=dev-python/pytest-xdist-1.18[${PYTHON_USEDEP}]
+		>=dev-python/py-1.5.2[${PYTHON_USEDEP}]
+		dev-python/six[${PYTHON_USEDEP}]
+		>=dev-python/virtualenv-16.0.0[${PYTHON_USEDEP}]
+	)
+"
+
+distutils_enable_tests pytest
+
+src_compile() {
+	local -x MYPY_USE_MYPYC=$(usex native-extensions 1 0)
+	distutils-r1_src_compile
+}
+
+python_test() {
+	local EPYTEST_DESELECT=(
+		# the majority of them require Internet (via pip)
+		mypy/test/testpep561.py
+	)
+	# Some mypy/test/testcmdline.py::PythonCmdlineSuite tests
+	# fail with high COLUMNS values
+	local -x COLUMNS=80
+
+	# The tests depend on having in-source compiled extensions if you want to
+	# test those compiled extensions. Copy them over *iff* they exist, and
+	# delete them afterwards.
+	(cd "${BUILD_DIR}/build/lib"*; find . -name '*.so' -exec cp {} "${S}"/{} \;)
```

Comment:
That's just… no. Ebuilds are not a place for hacks written on napkins dirty with ketchup.

Reply:
Do you have an alternative proposal to solve the fact that upstream's test suite only runs directly from the sources, including keeping their testsuite as an importable, installed subpackage which is incomplete and errors out because it relies on files that aren't installed?

Upstream's testsuite uses `pip install -e` run as a distinct stage before tox, with a suppression to stop tox from attempting to reinstall it. The editable install causes setuptools to do exactly what I did here -- place the compiled extensions in the source tree.

Reply:
Ebuilds are not a place for hacks written on napkins dirty with ketchup.

Reply:
- *.so is going to break on macOS at least - could use multilib.eclass' get_libname/get_modname, but
- I'd just copy the test data, it's not reliant on the layout, and it works on other platforms then too.

Reply:
Using get_modname would be incorrect and break on Darwin. The CPython build system uses .so on macOS. The situations where it doesn't use .so are:
- cygwin, uses .DLL
- Windows, uses .pyd
- HPUX, uses .sl

Reply:
I'm not sure if this is the case for us on Prefix at least:
```shell
$ ls -al usr/lib/python3.10/lib-dynload | head -5
total 4972
drwxr-xr-x  72 sam staff    2304 Jul 25  2022 .
drwxr-xr-x 205 sam staff    6560 Jul 25  2022 ..
-rwxr-xr-x   1 sam staff   69272 Jul 25  2022 _asyncio.cpython-310-darwin.bundle
-rwxr-xr-x   1 sam staff   18832 Jul 25  2022 _bisect.cpython-310-darwin.bundle
[...]
```

Reply:
That link doesn't seem to tell me much of anything at all, it's about libtool.

Python's own source code uses: https://github.com/python/cpython/blob/2209d814cab7630189faa0b9f9854986bd32f41c/Python/dynload_shlib.c#L37

See dynload_*.c for different platforms.

Depending on which _PyImport_DynLoadFiletab is compiled into python, the search pattern will be different. The technically-correct way to do this is to import importlib.machinery and iterate over members of EXTENSION_SUFFIXES.

If Gentoo Prefix is patching this then that means `pip install` inside of a virtualenv cannot install working wheels from PyPI. I hope in that case the platform identifier is sufficiently different to prevent pip from seeing those wheels as candidates to install.

Where's the python patch for this?

Reply:
Thanks, that's some excellent data on what's going on here. Not super exciting that it is, in fact, going on though.

The "good" news is that this oddity complicates Gentoo Prefix (and explains in its own patch that it's not a great idea, but refuses to rationalize making the change to begin with), but doesn't affect third-party consumers too much. It supports all four extensions! Fascinating(ly labyrinthine).

Begrudgingly, I accept that due to the existing hacks written on napkins dirty with ketchup, I can't use my ~~elegant bug for bug compatible implementation of setuptools' core logic~~ hacks written on napkins dirty with ketchup, and have to attempt to explore whether it is possible to use an even worse set of hacks written on napkins dirty with ketchup.

This is frustrating, since what I did was explicitly designed to be exactly what setuptools does in precisely this situation -- with the exception that running `python setup.py build_ext --inplace` would of course *first* recompile the whole lot, and we want to avoid the recompiles...